File: code/cpp/generate-parentheses.cc
// 3 ms, 09/26/2017
// Time  : O(n!) ?
// Space : O(n^2)
// EPI solution
class Solution {
public:
    vector<string> generateParenthesis(int n)
    {
        vector<string> result;
        gen_paren_recur("", n, n, &result, 0);
        return result;
    }
private:
    void gen_paren_recur(const string& s, int num_left_parens, int num_right_parens, vector<string>* res, int depth)
    {
        if (!num_right_parens) {
            res->emplace_back(s);
            return;
        }
        if (num_left_parens)
            gen_paren_recur(s + '(', num_left_parens - 1, num_right_parens, res, depth + 1);
        if (num_right_parens > num_left_parens)
            gen_paren_recur(s + ')', num_left_parens, num_right_parens - 1, res, depth + 1);
    }
};

File: code/python/generate-parentheses.py
# Date  : 03/20/2018
# Time  : O(c^n)
# Space : O(n)

class Solution:
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        res = []
        def gen_paren(s, left, right):
            if not right:
                res.append(s)
            if left:
                gen_paren(s + '(', left - 1, right)
            if right > left:
                gen_paren(s + ')', left, right - 1)
        gen_paren('', n, n)
        return res
